README.md
# FocusFlow for VSCode

![FocusFlow Logo](./icon.png)

FocusFlow is a productivity-boosting VSCode extension designed for developers who aim to optimize their coding sessions and manage tasks with precision. Seamlessly integrate the Pomodoro technique with an in-built Kanban board, right within your VSCode workspace.

## Features

1. **Pomodoro Timer:** Stay focused with integrated Pomodoro sessions. Easily start, pause, or stop without leaving your editor.
2. **Kanban Board:** Visual task management at your fingertips. Categorize tasks as 'To-Do', 'In Progress', and 'Done' to streamline your workflow.
3. **Seamless UI Integration:** Designed to effortlessly blend into the VSCode interface for an intuitive user experience.

## Usage

1. **Starting a Pomodoro Session:** Use the command palette (`Ctrl+Shift+P` or `Cmd+Shift+P`) and select "Start Pomodoro".
2. **Pausing/Stopping a Session:** Similarly, use the command palette to select "Pause Pomodoro" or "Stop Pomodoro".
3. **Managing Tasks:** Navigate to the FocusFlow Kanban in the Explorer sidebar to add, edit, or move tasks between columns.

## Local Setup

If you're looking to contribute or test the extension locally, follow these steps:

1. **Prerequisites:** Ensure you have Node.js installed.
   
2. **Clone the Repository:** 
   ```bash
   git clone https://github.com/your-repo-link/focusflow.git
   ```
3. **Navigate to Directory:**
   ```bash
   cd focusflow
   ```
4. **Install Dependencies:**
   ```bash
   npm install
   ```
5. **Compile:**
   ```bash
   npm run compile
   ```

Open in VSCode: Launch VSCode, and use the File -> Open Folder option to open the focusflow directory.

Run the Extension: Press F5 or navigate to the debugging section and click the "Run" button. This action will launch a new VSCode window with the FocusFlow extension loaded.


focusflow/CHANGELOG.md
# Change Log

All notable changes to the "focusflow" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

focusflow/src/extension.ts
import * as vscode from 'vscode';

class KanbanItem extends vscode.TreeItem {
    constructor(public readonly label: string, public collapsibleState: vscode.TreeItemCollapsibleState) {
        super(label, collapsibleState);
    }
}

class KanbanDataProvider implements vscode.TreeDataProvider<KanbanItem> {
    private _onDidChangeTreeData: vscode.EventEmitter<KanbanItem | undefined | null | void> = new vscode.EventEmitter<KanbanItem | undefined | null | void>();
    readonly onDidChangeTreeData: vscode.Event<KanbanItem | undefined | null | void> = this._onDidChangeTreeData.event;

    private tasks: string[] = [];
    private timer: string = "25:00";
    private timerRunning: boolean = false;

    startTimer() {
        this.timerRunning = true;
    }

    pauseTimer() {
        this.timerRunning = false;
    }

    stopTimer() {
        this.timerRunning = false;
        this.timer = "25:00";
        this._onDidChangeTreeData.fire();
    }

    constructor() {
        setInterval(() => {
            if (!this.timerRunning) return;

            const parts = this.timer.split(":");
            let minutes = parseInt(parts[0]);
            let seconds = parseInt(parts[1]);

            if (seconds > 0) {
                seconds -= 1;
            } else if (minutes > 0) {
                minutes -= 1;
                seconds = 59;
            }
            this.timer = `${minutes.toString().padStart(2, '0')}:${seconds.toString().padStart(2, '0')}`;
            this._onDidChangeTreeData.fire();
        }, 1000);
    }

    addTask(task: string) {
        this.tasks.push(task);
        this._onDidChangeTreeData.fire();
    }

    editTask(oldTask: string, newTask: string) {
        const index = this.tasks.indexOf(oldTask);
        if (index !== -1) {
            this.tasks[index] = newTask;
        }
        this._onDidChangeTreeData.fire();
    }

    removeTask(task: string) {
        const index = this.tasks.indexOf(task);
        if (index !== -1) {
            this.tasks.splice(index, 1);
        }
        this._onDidChangeTreeData.fire();
    }

    getTreeItem(element: KanbanItem): vscode.TreeItem {
        return element;
    }

    getChildren(element?: KanbanItem): Thenable<KanbanItem[]> {
        if (!element) {
            const timerItem = new KanbanItem(this.timer, vscode.TreeItemCollapsibleState.None);
            const taskItems = this.tasks.map(task => new KanbanItem(task, vscode.TreeItemCollapsibleState.None));
            return Promise.resolve([timerItem, ...taskItems]);
        }
        return Promise.resolve([]);
    }
}

export function activate(context: vscode.ExtensionContext) {
    const treeDataProvider = new KanbanDataProvider();
    vscode.window.registerTreeDataProvider('focusflow.kanban', treeDataProvider);

    vscode.commands.registerCommand('focusflow.addTask', async () => {
        const taskName = await vscode.window.showInputBox({ prompt: 'Enter the task name' });
        if (taskName) {
            treeDataProvider.addTask(taskName);
        }
    });

    vscode.commands.registerCommand('focusflow.editTask', async (item: KanbanItem) => {
        const newName = await vscode.window.showInputBox({ prompt: 'Edit the task name', value: item.label });
        if (newName) {
            treeDataProvider.editTask(item.label, newName);
        }
    });

    vscode.commands.registerCommand('focusflow.removeTask', (item: KanbanItem) => {
        treeDataProvider.removeTask(item.label);
    });

    vscode.commands.registerCommand('focusflow.startPomodoro', () => {
        treeDataProvider.startTimer();
    });

    vscode.commands.registerCommand('focusflow.pausePomodoro', () => {
        treeDataProvider.pauseTimer();
    });

    vscode.commands.registerCommand('focusflow.stopPomodoro', () => {
        treeDataProvider.stopTimer();
    });
}


focusflow/out/extension.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.activate = void 0;
const vscode = __importStar(require("vscode"));
class KanbanItem extends vscode.TreeItem {
    label;
    collapsibleState;
    constructor(label, collapsibleState) {
        super(label, collapsibleState);
        this.label = label;
        this.collapsibleState = collapsibleState;
    }
}
class KanbanDataProvider {
    _onDidChangeTreeData = new vscode.EventEmitter();
    onDidChangeTreeData = this._onDidChangeTreeData.event;
    tasks = [];
    timer = "25:00";
    timerRunning = false;
    startTimer() {
        this.timerRunning = true;
    }
    pauseTimer() {
        this.timerRunning = false;
    }
    stopTimer() {
        this.timerRunning = false;
        this.timer = "25:00";
        this._onDidChangeTreeData.fire();
    }
    constructor() {
        setInterval(() => {
            if (!this.timerRunning)
                return;
            const parts = this.timer.split(":");
            let minutes = parseInt(parts[0]);
            let seconds = parseInt(parts[1]);
            if (seconds > 0) {
                seconds -= 1;
            }
            else if (minutes > 0) {
                minutes -= 1;
                seconds = 59;
            }
            this.timer = `${minutes.toString().padStart(2, '0')}:${seconds.toString().padStart(2, '0')}`;
            this._onDidChangeTreeData.fire();
        }, 1000);
    }
    addTask(task) {
        this.tasks.push(task);
        this._onDidChangeTreeData.fire();
    }
    editTask(oldTask, newTask) {
        const index = this.tasks.indexOf(oldTask);
        if (index !== -1) {
            this.tasks[index] = newTask;
        }
        this._onDidChangeTreeData.fire();
    }
    removeTask(task) {
        const index = this.tasks.indexOf(task);
        if (index !== -1) {
            this.tasks.splice(index, 1);
        }
        this._onDidChangeTreeData.fire();
    }
    getTreeItem(element) {
        return element;
    }
    getChildren(element) {
        if (!element) {
            const timerItem = new KanbanItem(this.timer, vscode.TreeItemCollapsibleState.None);
            const taskItems = this.tasks.map(task => new KanbanItem(task, vscode.TreeItemCollapsibleState.None));
            return Promise.resolve([timerItem, ...taskItems]);
        }
        return Promise.resolve([]);
    }
}
function activate(context) {
    const treeDataProvider = new KanbanDataProvider();
    vscode.window.registerTreeDataProvider('focusflow.kanban', treeDataProvider);
    vscode.commands.registerCommand('focusflow.addTask', async () => {
        const taskName = await vscode.window.showInputBox({ prompt: 'Enter the task name' });
        if (taskName) {
            treeDataProvider.addTask(taskName);
        }
    });
    vscode.commands.registerCommand('focusflow.editTask', async (item) => {
        const newName = await vscode.window.showInputBox({ prompt: 'Edit the task name', value: item.label });
        if (newName) {
            treeDataProvider.editTask(item.label, newName);
        }
    });
    vscode.commands.registerCommand('focusflow.removeTask', (item) => {
        treeDataProvider.removeTask(item.label);
    });
    vscode.commands.registerCommand('focusflow.startPomodoro', () => {
        treeDataProvider.startTimer();
    });
    vscode.commands.registerCommand('focusflow.pausePomodoro', () => {
        treeDataProvider.pauseTimer();
    });
    vscode.commands.registerCommand('focusflow.stopPomodoro', () => {
        treeDataProvider.stopTimer();
    });
}
exports.activate = activate;
//# sourceMappingURL=extension.js.map